### PR TITLE
T4758: Rewrite show DHCP(v6) server leases  to vyos.opmode format

### PR DESCRIPTION
**Pull request by sever-sever**
## Change Summary

Rewrite op-mode DHCP and DHCPv6 leases to `vyos.opmode` format
Ability to show `raw`format

```
show dhcp server leases
show dhcpv6 server leases
```
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4758

## How to test
VyOS configuration:
```
vyos@r14:~$ show conf com | match dhcp
set service dhcp-server shared-network-name Lan01 authoritative
set service dhcp-server shared-network-name Lan01 name-server '1.1.1.1'
set service dhcp-server shared-network-name Lan01 subnet 192.0.2.0/24 default-router '192.0.2.1'
set service dhcp-server shared-network-name Lan01 subnet 192.0.2.0/24 lease '172800'
set service dhcp-server shared-network-name Lan01 subnet 192.0.2.0/24 range R1 start '192.0.2.10'
set service dhcp-server shared-network-name Lan01 subnet 192.0.2.0/24 range R1 stop '192.0.2.254'
set service dhcp-server shared-network-name Lan02 subnet 100.64.0.0/22 default-router '100.64.0.1'
set service dhcp-server shared-network-name Lan02 subnet 100.64.0.0/22 range R100 start '100.64.0.10'
set service dhcp-server shared-network-name Lan02 subnet 100.64.0.0/22 range R100 stop '100.64.0.254'
set service dhcp-server shared-network-name Lan02 subnet 100.64.0.0/22 range R200 start '100.64.1.20'
set service dhcp-server shared-network-name Lan02 subnet 100.64.0.0/22 range R200 stop '100.64.1.234'
set service dhcp-server shared-network-name Lan02 subnet 100.64.100.0/22 range Rpool100 start '100.64.100.5'
set service dhcp-server shared-network-name Lan02 subnet 100.64.100.0/22 range Rpool100 stop '100.64.100.55'
set service dhcpv6-server shared-network-name Lan-v6-02 subnet 2001:db8:23::/64 address-range start 2001:db8:23::199 stop '2001:db8:23::ffff'
set service dhcpv6-server shared-network-name Lan-v6-02 subnet 2001:db8:23::/64 name-server '2001:db8::ffff'

vyos@r14:~$ 
```

formatted output
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_server_leases --family inet
IP Address    Hardware address    State    Lease start          Lease expiration     Remaining        Pool    Hostname
------------  ------------------  -------  -------------------  -------------------  ---------------  ------  ----------
192.0.2.50    1a:f0:c4:ff:6a:2c   active   2022/10/31 16:39:27  2022/11/02 16:39:27  1 day, 22:49:20  Lan01   r1
192.0.2.51    32:e9:0a:d8:c9:21   active   2022/10/31 16:39:27  2022/11/02 16:39:27  1 day, 22:49:20  Lan01   r1
192.0.2.52    06:73:58:50:64:4e   active   2022/10/31 16:39:27  2022/11/02 16:39:27  1 day, 22:49:20  Lan01   r1
192.0.2.53    86:77:6f:7a:63:98   active   2022/10/31 16:39:27  2022/11/02 16:39:27  1 day, 22:49:20  Lan01   r1
192.0.2.54    a2:3f:8e:32:da:4a   active   2022/10/31 16:39:28  2022/11/02 16:39:28  1 day, 22:49:21  Lan01   r1
vyos@r14:~$ 
```

raw
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_server_leases --family inet --raw
[
    {
        "ip": "192.0.2.50",
        "state": "active",
        "pool": "Lan01",
        "end": 1667392767.0,
        "hardware": "1a:f0:c4:ff:6a:2c",
        "start": 1667219967.0,
        "hostname": "r1",
        "remaining": "1 day, 22:47:46"
    },
    {
        "ip": "192.0.2.51",
        "state": "active",
        "pool": "Lan01",
        "end": 1667392767.0,
        "hardware": "32:e9:0a:d8:c9:21",
        "start": 1667219967.0,
        "hostname": "r1",
        "remaining": "1 day, 22:47:46"
    },
    {
        "ip": "192.0.2.52",
        "state": "active",
        "pool": "Lan01",
        "end": 1667392767.0,
        "hardware": "06:73:58:50:64:4e",
        "start": 1667219967.0,
        "hostname": "r1",
        "remaining": "1 day, 22:47:46"
    },
    {
        "ip": "192.0.2.53",
        "state": "active",
        "pool": "Lan01",
        "end": 1667392767.0,
        "hardware": "86:77:6f:7a:63:98",
        "start": 1667219967.0,
        "hostname": "r1",
        "remaining": "1 day, 22:47:46"
    },
    {
        "ip": "192.0.2.54",
        "state": "active",
        "pool": "Lan01",
        "end": 1667392768.0,
        "hardware": "a2:3f:8e:32:da:4a",
        "start": 1667219968.0,
        "hostname": "r1",
        "remaining": "1 day, 22:47:47"
    }
]
vyos@r14:~$ 


```
Show pool statistics:
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_pool_statistics --family inet --raw
[
    {
        "pool": "Lan01",
        "size": 245,
        "leases": 5,
        "available": 240,
        "use_percentage": 2,
        "subnet": [
            "192.0.2.0/24"
        ]
    },
    {
        "pool": "Lan02",
        "size": 511,
        "leases": 0,
        "available": 511,
        "use_percentage": 0,
        "subnet": [
            "100.64.0.0/22",
            "100.64.100.0/22"
        ]
    }
]
vyos@r14:~$ 
```
IPv6
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_server_leases --family inet6 --raw
[
    {
        "ip": "2001:db8:23::95c2",
        "state": "active",
        "pool": "Lan-v6-02",
        "end": 1667263170.0,
        "last_communication": 1667219970.0,
        "iaid_duid": "00:00:00:00:00:04:50:4a:98:1d:6f:b3:49:ca:8a:43:c2:83:26:99:2b:d4",
        "type": "non-temporary",
        "remaining": "10:46:09"
    }
]
vyos@r14:~$ 

vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_pool_statistics --family inet6 --raw
[
    {
        "pool": "Lan-v6-02",
        "size": 65127,
        "leases": 1,
        "available": 65126,
        "use_percentage": 0,
        "subnet": [
            "2001:db8:23::/64"
        ]
    }
]
vyos@r14:~$ 
```

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

**Comment by sever-sever** [comment]
Only one bug that I found
It shows leases with same IP(v6) address
```
vyos@r14:~$ show dhcpv6  server leases 
IPv6 address       State    Last communication    Lease expiration     Remaining    Type           Pool       IAID_DUID
-----------------  -------  --------------------  -------------------  -----------  -------------  ---------  -----------------------------------------------------------------
2001:db8:23::95c2  active   2022/10/19 13:02:07   2022/10/20 01:02:07  7:18:32      non-temporary  Lan-v6-02  00:00:00:00:00:04:50:4a:98:1d:6f:b3:49:ca:8a:43:c2:83:26:99:2b:d4
2001:db8:23::95c2  active   2022/10/19 16:47:07   2022/10/20 04:47:06  11:03:31     non-temporary  Lan-v6-02  00:00:00:00:00:04:50:4a:98:1d:6f:b3:49:ca:8a:43:c2:83:26:99:2b:d4
vyos@r14:~$ 

```
Leases
```
vyos@r14:~$ cat /config/dhcpdv6.leases
# The format of this file is documented in the dhcpd.leases(5) manual page.
# This lease file was written by isc-dhcp-4.4.1

# authoring-byte-order entry is generated, DO NOT DELETE
authoring-byte-order little-endian;

server-duid "\000\001\000\001*\342\207sRT\000\3071\274";

ia-na "\000\000\000\000\000\004PJ\230\035o\263I\312\212C\302\203&\231+\324" {
  cltt 3 2022/10/19 10:02:07;
  iaaddr 2001:db8:23::95c2 {
    binding state active;
    preferred-life 27000;
    max-life 43200;
    ends 3 2022/10/19 22:02:07;
    set shared-networkname = "Lan-v6-02";
  }
}

ia-na "\000\000\000\000\000\004PJ\230\035o\263I\312\212C\302\203&\231+\324" {
  cltt 3 2022/10/19 13:47:07;
  iaaddr 2001:db8:23::95c2 {
    binding state active;
    preferred-life 26999;
    max-life 43199;
    ends 4 2022/10/20 01:47:06;
    set shared-networkname = "Lan-v6-02";
  }
}

vyos@r14:~$
```

Update, Fixed

**Review by jestabro — COMMENTED** [review]
For API, one should raise UnconfiguredSubsytem error:
```
diff --git a/src/op_mode/dhcp.py b/src/op_mode/dhcp.py
index c09dc693..67e5f0cc 100755
--- a/src/op_mode/dhcp.py
+++ b/src/op_mode/dhcp.py
@@ -130,14 +130,14 @@ def show_server_leases(raw: bool, family: str):
     if family == 'inet' and not config.exists('service dhcp-server'):
         unconf_message = 'DHCP server is not configured'
         if raw:
-            return {'error': True, 'reason': unconf_message}
+            raise vyos.opmode.UnconfiguredSubsystem(unconf_message)
         else:
             print(unconf_message)
         exit(0)
     if family == 'inet6' and not config.exists('service dhcpv6-server'):
         unconf_message = 'DHCPv6 server is not configured'
         if raw:
-            return {'error': True, 'reason': unconf_message}
+            raise vyos.opmode.UnconfiguredSubsystem(unconf_message)
         else:
             print(unconf_message)
         exit(0)
```

**Comment by sever-sever** [comment]
> For API, one should raise UnconfiguredSubsytem error:

Thanks, done!

File: src/op_mode/dhcp.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    leases = IscDhcpLeases(lease_file).get()
+    if pool is not None:
+        if config.exists(f'service dhcp-server shared-network-name {pool}'):
+            leases = list(filter(lambda x: _in_pool(x, pool), leases))
+    for lease in leases:
+        data_lease = {}
+        data_lease['ip'] = lease.ip
+        data_lease['state'] = lease.binding_state
+        data_lease['pool'] = lease.sets.get('shared-networkname', '')
+        data_lease['end'] = _utc_to_local(lease.end).strftime('%Y/%m/%d %H:%M:%S')
```
Comment (dmbaturin) [inline src/op_mode/dhcp.py]: I think we should only do timestamp to human-readable date conversion in human-readable output, but raw outputs should use timestamps.
If UIs and scripts want to do their own datetime display or calculation, sending timestamps in the reply will make their life a lot easier.

**Review by dmbaturin — CHANGES_REQUESTED** [review]

File: src/op_mode/dhcp.py (hunk 1 above)
Comment (sever-sever) [reply src/op_mode/dhcp.py]: I have an issue with the raw format and timestamp datetime
```
        #data_lease['end'] = _utc_to_local(lease.end).strftime('%Y/%m/%d %H:%M:%S')
        data_lease['end'] = datetime.strptime(str(lease.end), '%Y-%m-%d %H:%M:%S').timestamp()
```
Output:
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_server_leases --family inet
IP Address    Hardware address    State    Lease start          Lease expiration    Remaining    Pool    Hostname
------------  ------------------  -------  -------------------  ------------------  -----------  ------  ----------
192.0.2.47    ca:4f:54:b9:9a:27   active   2022-10-28 06:25:18  1.66701e+09         21:35:44     Lan01   r1
192.0.2.48    6a:2a:78:2c:6e:54   active   2022-10-28 06:25:18  1.66701e+09         21:35:44     Lan01   r1
192.0.2.49    4a:03:eb:1e:57:d5   active   2022-10-28 06:25:19  1.66701e+09         21:35:45     Lan01   r1
192.0.2.46    9a:07:79:19:72:b3   active   2022-10-28 06:25:21  1.66701e+09         21:35:47     Lan01   r1
192.0.2.45    fa:ca:03:00:35:a7   active   2022-10-28 06:25:22  1.66701e+09         21:35:48     Lan01   r1
vyos@r14:~$ 
```

Raw:
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_server_leases --family inet --raw
Traceback (most recent call last):
  File "/usr/libexec/vyos/op_mode/dhcp.py", line 276, in <module>
    res = vyos.opmode.run(sys.modules[__name__])
  File "/usr/lib/python3/dist-packages/vyos/opmode.py", line 152, in run
    return dumps(res, indent=4)
  File "/usr/lib/python3.9/json/__init__.py", line 234, in dumps
    return cls(
  File "/usr/lib/python3.9/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/lib/python3.9/json/encoder.py", line 429, in _iterencode
    yield from _iterencode_list(o, _current_indent_level)
  File "/usr/lib/python3.9/json/encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "/usr/lib/python3.9/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/lib/python3.9/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
vyos@r14:~$ 
```

Comment (sever-sever) [reply src/op_mode/dhcp.py]: in Progress...

Comment (sever-sever) [reply src/op_mode/dhcp.py]: Done!

**Review by dmbaturin — APPROVED** [review]

**Review by dmbaturin — APPROVED** [review]

**Review by c-po — APPROVED** [review]

**Pull request merged**